lighthouse-core/scripts/i18n/bake-strings.js

```diff
@@ -0,0 +1,144 @@
+#!/usr/bin/env node
+/**
+ * @license Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+'use strict';
+
+/* eslint-disable no-console, max-len */
+
+const fs = require('fs');
+const path = require('path');
+
+const LH_ROOT = path.join(__dirname, '../../../');
+
+/**
+ * @typedef ICUMessageDefn
+ * @property {string} message
+ * @property {string} [description]
+ * @property {string} [meaning]
+ * @property {Record<string, ICUPlaceholderDefn>} [placeholders]
+ */
+
+/**
+ * @typedef ICUPlaceholderDefn
+ * @property {string} content
+ * @property {string} [example]
+ */
+
+const ignoredPathComponents = [
+  '/.git',
+  '/scripts',
+  '/node_modules',
+  '/test/',
+  '-test.js',
+  '-renderer.js',
+];
+
+/**
+ * Take a series of CTC format ICU messages and converts them to LHL format by
+ * replacing $placeholders$ with their {ICU} values. Functional opposite of
+ * `convertMessageToPlaceholders`. This is commonly called as the last step in
+ * translation.
+ *
+ * Converts this:
+ * messages: {
+ *  "lighthouse-core/audits/seo/canonical.js | explanationDifferentDomain" {
+ *    "message": "Points to a different domain ($ICU_0$)",
+ *    "placeholders": {
+ *      "ICU_0": {
+ *        "content": "{url}",
+ *        "example": "https://example.com/"
+ *      },
+ *    },
+ *  },
+ * }
+ *
+ * Into this:
+ * messages: {
+ *  "lighthouse-core/audits/seo/canonical.js | explanationDifferentDomain" {
+ *    "message": "Points to a different domain ({url})",
+ *    },
+ *  },
+ * }
+ *
+ * Throws if there is a $placeholder$ in the message that has no corresponding
+ * value in the placeholders object, or vice versa.
+ *
+ * @param {Record<string, ICUMessageDefn>} messages
+ * @returns {Record<string, ICUMessageDefn>}
+ */
+function bakePlaceholders(messages) {
+  for (const [_, defn] of Object.entries(messages)) {
+    // Don't need descriptions or meanings anymore.
+    delete defn['description'];
+    delete defn['meaning'];
+
+    let message = defn['message'];
+    const placeholders = defn['placeholders'];
+
+    for (const placeholder in placeholders) {
+      if (!Object.prototype.hasOwnProperty.call(placeholders, placeholder)) continue;
+
+      const content = placeholders[placeholder]['content'];
+      if (!message.includes(`$${placeholder}$`)) {
+        throw Error(`Message "${message}" has extra placeholder "${placeholder}"`);
```

Comment:
kind of confusing...maybe `Extra placeholder "${placeholder}" not found in message "${message}"` or something?